### ISSUE TITLE
Merging into beta is not detected as a prerelease

### ISSUE THREAD

**Issue by philipp-rudloff-sendcloud** [semantic-release/semantic-release#1995]
## Current behavior

While working on a merge request, I noticed that the `test:release` job which runs the semantic-release script in dry run mode is not detecting the fact that the release should be a prerelease as the MR is targeting the project’s `beta` branch. Everything in the output is fully as expected save for the following two parts:

- [11:56:41 AM] [semantic-release] › ℹ  The next release version is 2.0.0
- [11:56:42 AM] [semantic-release] › ✔  Published release 2.0.0 on default channel

The merge request targets the `beta` branch and the semantic release configuration does not set the `branches` option and should therefor default to the values listed in https://semantic-release.gitbook.io/semantic-release/usage/configuration#branches.

I can unfortunately not share links to a public repository because the project in question is a private work project. I tried my best to include a cleaned-up version of the CI job output that still contains all necessary troubleshooting information but no work specifics. Let me know if you need other information.

## Expected behavior

The next release version is detected as 2.0.0-beta.1 because the target branch of the MR is `beta` and no beta release on the major version 2 exists, yet. The most recent beta version is 1.0.0-beta.10 and the most recent regular version is 1.2.0.

## Environment

- **semantic-release** version: 17.4.4
- CI environment: GitLab CI/CD
- Plugins used: @semantic-release/changelog, @semantic-release/commit-analyzer, @semantic-release/git, @semantic-release/npm, @semantic-release/release-notes-generator
- **semantic-release** configuration: release.config.js:

  ```js
  module.exports = {
    plugins: [
      '@semantic-release/commit-analyzer',
      '@semantic-release/release-notes-generator',
      '@semantic-release/changelog',
      '@semantic-release/git',
      '@semantic-release/npm',
    ],
    verifyConditions: ['@semantic-release/changelog', '@semantic-release/git', '@semantic-release/npm'],
    prepare: ['@semantic-release/changelog', '@semantic-release/npm', '@semantic-release/git'],
    publish: ['@semantic-release/npm'],
  }
  ```

<details>
<summary>Shortened semantic-release output from a CI job</summary>

```
$ npm run release:dry-run -- --branches=$CI_COMMIT_REF_NAME --no-ci
> @organization/package-name@1.2.0 release:dry-run /path/to/projects/package-name
> npm run semantic-release -- --dry-run --debug "--branches=my-branch-name" "--no-ci"
> @organization/package-name@1.2.0 presemantic-release
> npm run build:rollup
> @organization/package-name@1.2.0 prebuild:rollup
> npm run clean
> @organization/package-name@1.2.0 clean
> rimraf dist public
> @organization/package-name@1.2.0 build:rollup
> cross-env NODE_ENV=production rollup --config
src/package-name.js → dist/package-name.js...
created dist/package-name.js in 4.3s
src/package-name.css → dist/package-name.css...
(!) The emitted file "package-name.css" overwrites a previously emitted file of the same name.
created dist/package-name.css in 2.3s
> @organization/package-name@1.2.0 semantic-release
> cross-env GIT_AUTHOR_NAME=Author GIT_AUTHOR_EMAIL=author@example.com GIT_COMMITTER_NAME=Author GIT_COMMITTER_EMAIL=author@example.com semantic-release "--dry-run" "--debug" "--branches=my-branch-name" "--no-ci"
[11:56:33 AM] [semantic-release] › ℹ  Running semantic-release version 17.4.4
2021-06-28T11:56:33.723Z semantic-release:config load config from: /path/to/projects/package-name/release.config.js
2021-06-28T11:56:33.727Z semantic-release:config options values: {
  branches: [ 'my-branch-name' ],
  repositoryUrl: 'https://gitlab.com/organization/package-name',
  tagFormat: 'v${version}',
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    '@semantic-release/changelog',
    '@semantic-release/git',
    '@semantic-release/npm'
  ],
  verifyConditions: [
    '@semantic-release/changelog',
    '@semantic-release/git',
    '@semantic-release/npm'
  ],
  prepare: [
    '@semantic-release/changelog',
    '@semantic-release/npm',
    '@semantic-release/git'
  ],
  publish: [ '@semantic-release/npm' ],
  _: [],
  'dry-run': true,
  d: true,
  dryRun: true,
  debug: true,
  b: [ 'my-branch-name' ],
  ci: false,
  '$0': 'node_modules/.bin/semantic-release',
  noCi: true
}
2021-06-28T11:56:33.968Z semantic-release:plugins options for @semantic-release/changelog/verifyConditions: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/changelog"
2021-06-28T11:56:33.970Z semantic-release:plugins options for @semantic-release/git/verifyConditions: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/git"
2021-06-28T11:56:33.970Z semantic-release:plugins options for @semantic-release/npm/verifyConditions: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/npm"
2021-06-28T11:56:33.971Z semantic-release:plugins options for @semantic-release/commit-analyzer/analyzeCommits: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
2021-06-28T11:56:33.971Z semantic-release:plugins options for @semantic-release/release-notes-generator/generateNotes: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
2021-06-28T11:56:33.972Z semantic-release:plugins options for @semantic-release/changelog/prepare: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/changelog"
2021-06-28T11:56:33.972Z semantic-release:plugins options for @semantic-release/npm/prepare: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/npm"
2021-06-28T11:56:33.972Z semantic-release:plugins options for @semantic-release/git/prepare: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/git"
2021-06-28T11:56:33.973Z semantic-release:plugins options for @semantic-release/npm/publish: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/npm"
2021-06-28T11:56:33.973Z semantic-release:plugins options for @semantic-release/npm/addChannel: {}
[11:56:33 AM] [semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/npm"
2021-06-28T11:56:34.004Z semantic-release:get-git-auth-url Verifying ssh auth by attempting to push to  https://gitlab.com/organization/package-name
2021-06-28T11:56:34.932Z semantic-release:git Error: Command failed with exit code 128: git push --dry-run --no-verify https://gitlab.com/organization/package-name HEAD:my-branch-name
remote: HTTP Basic: Access denied
fatal: Authentication failed for 'https://gitlab.com/organization/package-name.git/'
    at makeError (/path/to/projects/package-name/node_modules/semantic-release/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/path/to/projects/package-name/node_modules/semantic-release/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async verifyAuth (/path/to/projects/package-name/node_modules/semantic-release/lib/git.js:207:5)
    at async module.exports (/path/to/projects/package-name/node_modules/semantic-release/lib/get-git-auth-url.js:91:5)
    at async run (/path/to/projects/package-name/node_modules/semantic-release/index.js:56:27)
    at async module.exports (/path/to/projects/package-name/node_modules/semantic-release/index.js:260:22)
    at async module.exports (/path/to/projects/package-name/node_modules/semantic-release/cli.js:55:5)
2021-06-28T11:56:34.933Z semantic-release:get-git-auth-url SSH key auth failed, falling back to https.
2021-06-28T11:56:40.337Z semantic-release:get-tags found tags for branch my-branch-name: [ { gitTag: 'v1.0.0', version: '1.0.0', channels: [ null ] }, { gitTag: 'v1.0.0-beta.1', version: '1.0.0-beta.1', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.10', version: '1.0.0-beta.10', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.2', version: '1.0.0-beta.2', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.3', version: '1.0.0-beta.3', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.4', version: '1.0.0-beta.4', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.5', version: '1.0.0-beta.5', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.6', version: '1.0.0-beta.6', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.7', version: '1.0.0-beta.7', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.8', version: '1.0.0-beta.8', channels: [ 'beta' ] }, { gitTag: 'v1.0.0-beta.9', version: '1.0.0-beta.9', channels: [ 'beta' ] }, { gitTag: 'v1.0.1', version: '1.0.1', channels: [ null ] }, { gitTag: 'v1.0.2', version: '1.0.2', channels: [ null ] }, { gitTag: 'v1.0.3', version: '1.0.3', channels: [ null ] }, { gitTag: 'v1.1.0', version: '1.1.0', channels: [ null ] }, { gitTag: 'v1.1.1', version: '1.1.1', channels: [ null ] }, { gitTag: 'v1.1.2', version: '1.1.2', channels: [ null ] }, { gitTag: 'v1.1.3', version: '1.1.3', channels: [ null ] }, { gitTag: 'v1.1.4', version: '1.1.4', channels: [ null ] }, { gitTag: 'v1.2.0', version: '1.2.0', channels: [ null ] } ]
[11:56:40 AM] [semantic-release] › ⚠  Run automated release from branch my-branch-name on repository https://gitlab-ci-token:[secure]@gitlab.com/organization/package-name in dry-run mode
[11:56:41 AM] [semantic-release] › ✔  Allowed to push to the Git repository
[11:56:41 AM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/changelog"
[11:56:41 AM] [semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/changelog"
[11:56:41 AM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/git"
[11:56:41 AM] [semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/git"
[11:56:41 AM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/npm"
[11:56:41 AM] [semantic-release] [@semantic-release/npm] › ℹ  Verify authentication for registry https://registry.npmjs.org/
[11:56:41 AM] [semantic-release] [@semantic-release/npm] › ℹ  Wrote NPM_TOKEN to /tmp/bla/.npmrc
username
[11:56:41 AM] [semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/npm"
[11:56:41 AM] [semantic-release] › ℹ  Found git tag v1.2.0 associated with version 1.2.0 on branch my-branch-name
2021-06-28T11:56:41.813Z semantic-release:get-commits Use from: 55bfbfea9a1c236af54032e84db88f5a56341133
[11:56:41 AM] [semantic-release] › ℹ  Found 1 commits since last release
2021-06-28T11:56:41.828Z semantic-release:get-commits Parsed commits: [omitted]
[11:56:41 AM] [semantic-release] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[11:56:41 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: feat: title
BREAKING CHANGE: Message.
2021-06-28T11:56:41.943Z semantic-release:commit-analyzer Analyzing with default rules
2021-06-28T11:56:41.944Z semantic-release:commit-analyzer The rule { breaking: true, release: 'major' } match commit with release type 'major'
2021-06-28T11:56:41.945Z semantic-release:commit-analyzer Release type 'major' is the highest possible. Stop analysis.
[11:56:41 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The release type for the commit is major
[11:56:41 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analysis of 1 commits complete: major release
[11:56:41 AM] [semantic-release] › ✔  Completed step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[11:56:41 AM] [semantic-release] › ℹ  The next release version is 2.0.0
[11:56:41 AM] [semantic-release] › ℹ  Start step "generateNotes" of plugin "@semantic-release/release-notes-generator"
2021-06-28T11:56:42.133Z semantic-release:release-notes-generator version: '2.0.0'
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator host: undefined
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator owner: 'organization'
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator repository: 'package-name'
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator previousTag: 'v1.2.0'
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator currentTag: 'v2.0.0'
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator host: 'https://gitlab.com'
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator linkReferences: undefined
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator issue: 'issues'
2021-06-28T11:56:42.134Z semantic-release:release-notes-generator commit: 'commit'
[11:56:42 AM] [semantic-release] › ✔  Completed step "generateNotes" of plugin "@semantic-release/release-notes-generator"
[11:56:42 AM] [semantic-release] › ⚠  Skip step "prepare" of plugin "@semantic-release/changelog" in dry-run mode
[11:56:42 AM] [semantic-release] › ⚠  Skip step "prepare" of plugin "@semantic-release/npm" in dry-run mode
[11:56:42 AM] [semantic-release] › ⚠  Skip step "prepare" of plugin "@semantic-release/git" in dry-run mode
[11:56:42 AM] [semantic-release] › ⚠  Skip v2.0.0 tag creation in dry-run mode
[11:56:42 AM] [semantic-release] › ⚠  Skip step "publish" of plugin "@semantic-release/npm" in dry-run mode
[11:56:42 AM] [semantic-release] › ✔  Published release 2.0.0 on default channel
[11:56:42 AM] [semantic-release] › ℹ  Release note for version 2.0.0:
# 2.0.0 (https://gitlab.com/organization/package-name/compare/v1.2.0...v2.0.0) (2021-06-28)
### Features
    * title (e72a4e2 (https://gitlab.com/organization/package-name/commit/e72a4e23cd5c69b627fb61339bf025f69542dde8))
### BREAKING CHANGES
    * Message.
```
    
</details>

(Note: I thought at first that the logged option values at the top would include the branches option as detected from the configuration, but that isn’t actually the case.)

**Comment by philipp-rudloff-sendcloud** [comment]
I think I just figured out why this is happening. While running the CI jobs on my merge request’s branch, the branch that semantic-release looks at is the MR’s branch, not `beta`. Only once I merge into `beta` would the release script actually behave as expected.

Is there a way for me to tell semantic-release to behave as if it’s running on a different branch so I can keep a reliable dry run of semantic-release in the CI pipelines that actually reflects what would happen if I were to press the big ol’ merge button? That would be really helpful to gain confidence in the release process going smoothly.

**Comment by philipp-rudloff-sendcloud** [comment]
> Is there a way for me to tell semantic-release to behave as if it’s running on a different branch so I can keep a reliable dry run of semantic-release in the CI pipelines that actually reflects what would happen if I were to press the big ol’ merge button?

This can be done by providing the `branches` option with the target branch instead of the branch of the MR like so:

```
test:release:
  extends: .test_template
  script:
    - |
      if [ ! -z "$CI_MERGE_REQUEST_TARGET_BRANCH_NAME" ]; then
        semantic-release --dry-run --branches=$CI_MERGE_REQUEST_TARGET_BRANCH_NAME --no-ci
      else
        semantic-release --dry-run --branches=$CI_COMMIT_REF_NAME --no-ci
      fi
```

**Comment by philipp-rudloff-sendcloud** [comment]
> Is there a way for me to tell semantic-release to behave as if it’s running on a different branch so I can keep a reliable dry run of semantic-release in the CI pipelines that actually reflects what would happen if I were to press the big ol’ merge button? That would be really helpful to gain confidence in the release process going smoothly.

I filed a feature request for this here: https://github.com/semantic-release/semantic-release/issues/2015#issue-938827189
